pkg/registry/flowcontrol/ensurer/strategy.go

```diff
@@ -215,130 +208,130 @@ func setAutoUpdateAnnotation(accessor metav1.Object, autoUpdate bool) {
 	}
 
-	klog.V(5).InfoS(fmt.Sprintf("The %s already exists, checking whether it is up to date", wrapper.TypeName()), "type", configurationType, "name", name)
-	newObject, update, err := strategy.ShouldUpdate(current, bootstrap)
+	klog.V(5).InfoS(fmt.Sprintf("The %s already exists, checking whether it is up to date", bootstrap.typeName()), "type", configurationType, "name", name)
+	newObject, update, err := strategy.ShouldUpdate(wah)
 	if err != nil {
-		return fmt.Errorf("failed to determine whether auto-update is required for %s type=%s name=%q error=%w", wrapper.TypeName(), configurationType, name, err)
+		return fmt.Errorf("failed to determine whether auto-update is required for %s type=%s name=%q error=%w", bootstrap.typeName(), configurationType, name, err)
 	}
 	if !update {
 		if klogV := klog.V(5); klogV.Enabled() {
-			klogV.InfoS("No update required", "wrapper", wrapper.TypeName(), "type", configurationType, "name", name,
-				"diff", cmp.Diff(current, bootstrap))
+			klogV.InfoS("No update required", "wrapper", bootstrap.typeName(), "type", configurationType, "name", name,
+				"diff", cmp.Diff(wah.getHave(), wah.getWant()))
 		}
 		return nil
 	}
 
-	if _, err = wrapper.Update(newObject); err == nil {
-		klog.V(2).Infof("Updated the %s type=%s name=%q diff: %s", wrapper.TypeName(), configurationType, name, cmp.Diff(current, newObject))
+	if err = newObject.update(ctx); err == nil {
+		klog.V(2).Infof("Updated the %s type=%s name=%q diff: %s", bootstrap.typeName(), configurationType, name, cmp.Diff(wah.getHave(), wah.getWant()))
 		return nil
 	}
 
 	if apierrors.IsConflict(err) {
-		klog.V(2).InfoS(fmt.Sprintf("Something updated the %s concurrently, I will check its spec later", wrapper.TypeName()), "type", configurationType, "name", name)
+		klog.V(2).InfoS(fmt.Sprintf("Something updated the %s concurrently, I will check its spec later", bootstrap.typeName()), "type", configurationType, "name", name)
 		return nil
 	}
 
-	return fmt.Errorf("failed to update the %s, will retry later type=%s name=%q error=%w", wrapper.TypeName(), configurationType, name, err)
+	return fmt.Errorf("failed to update the %s, will retry later type=%s name=%q error=%w", bootstrap.typeName(), configurationType, name, err)
 }
 
-// removeAutoUpdateEnabledConfiguration makes an attempt to remove the given
-// configuration object if automatic update of the spec is enabled for this object.
-func removeAutoUpdateEnabledConfiguration(wrapper configurationWrapper, name string) error {
-	current, err := wrapper.Get(name)
+// RemoveUnwantedObjects attempts to delete the configuration objects
+// that exist, are annotated `apf.kubernetes.io/autoupdate-spec=true`, and do not
+// have a name in the given set.  A refusal due to concurrent update is logged
+// and not considered an error; the object will be reconsidered later.
+func RemoveUnwantedObjects(ctx context.Context, boots BootstrapObjects) error {
+	current, err := boots.getExistingObjects()
```

Comment:
Hmm - why you return nil here?
I would suggest returning the same error that we were returning before:
```
return fmt.Errorf("failed to retrieve the %s, will retry later name=%q error=%w", wrapper.TypeName(), name, err)
```

This error is bubbles up to `ensure()` method so it matter in two places:
https://github.com/kubernetes/kubernetes/pull/111422/files#diff-2db4d80439afc0612e686dfe2939c42ff1e620997a123d4fd5453ecbf7cec009R158
https://github.com/kubernetes/kubernetes/pull/111422/files#diff-2db4d80439afc0612e686dfe2939c42ff1e620997a123d4fd5453ecbf7cec009R180

In the first one, I would say that retrying is desired.
In the second, it only adds an error log, which also seems desired.

Reply:
Oh crud!  That's a typo.  I meant to `return err`.  Fixed in the latest rev.